Fixme estimateScriptsCost

v0d1ch opened the pull request:
- Do not throw from a pure function
- Introduce new constructor ErrTranslationError
- Remove BadTranslationException
- remove the FIXME

ffakenz committed:
Add failing test

ffakenz committed:
Refactor to render feedbacks log enough for the client to consume.

Refactor State to have a list of feedbacks instead of a maybe value.
This list will only contain `non expired` feedbacks.
A feedback is expired if it is 3 seconds older than current State time.

This refactor means now the widget to display is not a plain str anymore
but a list str widgets.

Note: No feedback will be render unless current State time is ahead of
TUI application start up time.

Extra: Added shouldNotRender helper to check feedbacks are getting clear.

v0d1ch committed:
Fixme estimateScriptsCost

- Do not throw from a pure function
- Introduce new constructor ErrTranslationError
- Remove BadTranslationException
- remove the FIXME

ffakenz reviewed (approved):
Looks good but are there no tests affected by this change? 🤔
If not, do you think it worth to add them?

v0d1ch commented:
> Looks good but are there no tests affected by this change? thinking If not, do you think it worth to add them?

I guess I could introduce a new test case for `estimateScriptsCost`  so that I can trigger the both possible errors from that function.

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/TxSpec.hs:
```diff
@@ -179,6 +179,7 @@ spec =
                                           ErrNotEnoughFunds{} -> "Not enough funds"
                                           ErrUnknownInput{} -> "Unknown input"
                                           ErrScriptExecutionFailed{} -> "Script(s) execution failed"
+                                          ErrTranslationError{} -> "Transaction context translation error"
```
Formatting loons off.. are you using fourmolu already?

v0d1ch replied:
I configured it and I can see it in action. I'll show you in the meeting but for this line I don't see what is off?

ch1bo replied:
Nevermind, it was only looking weird on mobile.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Wallet.hs:
```diff
@@ -361,15 +360,13 @@ estimateScriptsCost ::
   case result of
     Left translationError ->
-      throw $ BadTranslationException translationError
+      Left $ ErrTranslationError translationError
```
This is better. However, `ErrCoverFee` is a bit of a kitchen sink. The next step would be to have a dedicated error type for this function or finally bite the bullet to refactor the whole module to properly use cardano-api types.

ch1bo reviewed (approved):
Change looks good and aligns nicely with existing error types. They become a bit of a kitchen sink now, but nothing to worry for this PR.

ffakenz committed:
Add playground to init investigation.

Added traces to seedFaucet for debugging.
Most of this code will be removed once the issue is fixed.

By increasing the amount of threads to a 100 we observed another
Node.Socket exception.
This means querying the `cardano-node` concurrently may fail for many
reasons and we must handle these exceptions propperly.

ffakenz committed:
Improve traces and reduce nbr of threads running concurrently

ffakenz committed:
Always show tracing after test finishes (with success or error)

ch1bo committed:
Checkout full history when building the docs

We recently added last updated/translated at labels and these use the
git history to determine the dates.

ffakenz committed:
Fix feedback naming

ffakenz committed:
Improve feedback report function to use the lens function  instead

ffakenz committed:
Fix error feedback msg on  helper

ffakenz committed:
Remove tracer from seedFromFaucet and extend the retry on submit
to handle Node.Socket.connect ResourceExhausted exception along with
SubmitTransactionException

ffakenz committed:
Remove unnecessary hlint ignore

ffakenz committed:
Remove unnecessary ghc option

ch1bo committed:
Add a page for API Behavior

This is something which was mentioned to be helpful by the internal
auditor. Not sure whether we want to keep it here, move to the haddock
or the api-reference source / api.yaml.

ch1bo committed:
Move ClientInput/ServerOutput into Hydra.API

When referring to these modules/datatypes it occurred to me that putting
them into the `Hydra.API` scope makes it clear they are part of that layer.

ffakenz committed:
Fix seedFromFaucet to retry in case of expected exceptions

Expected exceptions are:
- `SubmitTransactionException`, comming from `submitTransaction`
- `IOException.ResourceExhausted`, commig from `Node.Socket.connect`

Add tracer back to seedFrom faucet to log when expected exception
occurs.

Other `IOExceptions` will be thrown causing the spec to fail.

abailly committed:
Timeout after 10s waiting for UTxO in wallet

ffakenz committed:
Minor fix on e2e spec

ch1bo commented on hydra-node/test/Hydra/Chain/Direct/WalletSpec.hs:
```diff
@@ -232,6 +236,30 @@ prop_removeUsedInputs =
+    forAllBlind (reasonablySized $ genOutputsForInputs tx) $ \txUTxO ->
+      let
+          -- TODO: update the tx here so that we can see the test failing
+          -- generated tx does not contain any script data
```
It was approved.. now it requires changes again to address this TODO.

(I don't think these tests are worth their weight. Especially as I see a more holistic rewrite of the Wallet upcoming)

v0d1ch replied:
Ok, I'll remove them from this pr then. Do you agree @ffakenz ?

ffakenz replied:
I would not remove the test, especially knowing we are going to refactor Wallet (i think they will come in handy).

But I don't have a strong opinion. Would leave it for the rest to decide on this 🙏

v0d1ch replied:
The test that was supposed to fail `prop_failsWithScriptExecutionFailed` was actually not implemented. And `prop_correctTxIsNotErroringOut` is green because the tx gen we are testing with is not actually using any scripts so `estimateScriptsCost` always produces `Right`. I'll nuke it for now.

ffakenz committed:
Remove redundant exception handle

ffakenz committed:
Do not trace submit tx exceptions

ffakenz committed:
Add FaucetLog

ffakenz committed:
Refactor draw feedback to only render latest log with time without
clearing the state on every event handle.

ch1bo committed:
Revert "Merge pull request #493 from input-output-hk/abailly-iohk/fix-direct-chain-flakiness"

This reverts commit 97a5650cd24b264837ddf7c0ea233ace7f7d55e0, reversing
changes made to b0ebec965f27553a1c99e5457441219dcb9306d6.

ffakenz committed:
Fix spec as not we expect not to clear the feedback anymore

ffakenz committed:
Minor fix on tui spec

ffakenz committed:
change separator to draw feedback time and msg

Co-authored-by: Sebastian Nagel <ch1bo@users.noreply.github.com>

ffakenz committed:
Remove unnecesary splittable instance which was used to make the list widget to work

Remove unnecesary hiding imports

ffakenz committed:
Refactor exception handling aesthetics
